src/backend/distributed/utils/resource_lock.c

```diff
@@ -360,11 +464,33 @@ LockShardsInPlacementListMetadata(List *shardPlacementList, LOCKMODE lockMode)
 void
 LockShardListResources(List *shardIntervalList, LOCKMODE lockMode)
 {
+	List *referenceShardIntervalList = NIL;
 	ListCell *shardIntervalCell = NULL;
 
 	/* lock shards in order of shard id to prevent deadlock */
 	shardIntervalList = SortList(shardIntervalList, CompareShardIntervalsById);
 
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		int64 shardId = shardInterval->shardId;
+
+		/*
+		 * If it is a reference table shard, lock should be get on worker nodes
+		 * to prevent divergence of shard placement on different worker nodes.
+		 */
+		if (ReferenceTableShardId(shardId) && QueryRunOnMXCluster())
```

Comment:
I really don't want to add code complexity to locking functions. They are currently very simple, and lets try to keep them as simple.

I can see that you're first trying to get all the reference table shards that needs to locked on the remote node. That makes sense, but, can we move the logic to a seperate function like:
```C
LockShardResroucesOnFirstWorkerIfNecessary()
     foreach shardId:
     if (ReferenceTableShardId(shardId) && QueryRunOnMXCluster())
          LockShardListResourcesOnFirstWorker(lockMode, list_make1(shardInterval));
```

Later, change here and `AcquireExecutorLocks` to something like:
```C
   if (LockShardResroucesOnFirstWorkerIfNecessary() != SHARD_RESOURCES_LOCKED_LOCALLY)
        LockShardResource()
```